lib/brush.cpp

```diff
@@ -0,0 +1,91 @@
+#include "brush.h"
+
+#include <QPainter>
+#include <QtMath>
+#include <QSet>
+
+Brush::Brush(const Brush::Type type, const QSize &size, const QPointF handle, const bool relativeHandle) :
+    type(type), size(size), pixmap(), handle(relativeHandle ? QPointF(handle.x() * qreal(size.width()), handle.y() * qreal(size.height())) : handle)
+{
+    if (type == ImageType) {
+        pixmap = QPixmap::fromImage(QImage(size, QImage::Format_ARGB32));
+    }
+    else {
+        QImage image = QImage(size, QImage::Format_ARGB32);
+        image.fill(qRgba(0, 0, 0, 0));
+        const QRgb colour = qRgba(255, 255, 255, 255);
+        if (type == SquareType) {
+            fillRectangle(image, image.rect(), image.rect(), colour);
+        }
+        else {
+            fillEllipse(image, image.rect(), image.rect(), colour);
+        }
+        pixmap = QBitmap::fromImage(image.createMaskFromColor(colour, Qt::MaskOutColor));
+    }
+}
+
+inline void Brush::drawPixel(QImage &image, const QRect &clip, const QPoint point, const QRgb colour)
+{
+    Q_ASSERT(image.format() == QImage::Format_ARGB32);
+
+    const QRect clipped = clip & image.rect();
+    if (clipped.contains(point))
+        *(reinterpret_cast<QRgb *>(image.scanLine(point.y())) + point.x()) = colour;
+}
+
+inline void Brush::drawSpan(QImage &image, const QRect &clip, const int x0, const int x1, const int y, const QRgb colour) {
```

Comment:
Opening curly brace for functions should be on its own line. I mostly follow Qt's coding style out of habit:

https://wiki.qt.io/Qt_Coding_Style